File: app/src/main/java/com/bookyrself/bookyrself/services/FirebaseService.java
package com.bookyrself.bookyrself.services;

import com.bookyrself.bookyrself.models.SerializedModels.EventCreationResponse;
import com.bookyrself.bookyrself.models.SerializedModels.EventDetail.EventDetail;
import com.bookyrself.bookyrself.models.SerializedModels.User.EventInviteInfo;
import com.bookyrself.bookyrself.models.SerializedModels.User.User;

import java.util.HashMap;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

/**
 * Created by benmedcalf on 11/22/17.
 */

public class FirebaseService {

    private static FirebaseService.FirebaseApi INSTANCE;

    private static String BASE_URL_BOOKYRSELF_FIREBASE = "https://bookyrself-staging.firebaseio.com/";

    public static FirebaseService.FirebaseApi getAPI() {

        if (INSTANCE == null) {
            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            OkHttpClient client = new OkHttpClient.Builder().addInterceptor(interceptor).build();


            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL_BOOKYRSELF_FIREBASE)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                    .build();
            INSTANCE = retrofit.create(FirebaseService.FirebaseApi.class);
        }
        return INSTANCE;
    }

    public interface FirebaseApi {
        @GET("/events/{id}.json")
        Call<EventDetail> getEventData(@Path("id") String eventId);

        @GET("/users/{id}/events.json")
        Call<HashMap<String, EventInviteInfo>> getUsersEventInfo(@Path("id") String userId);

        @GET("/users/{id}/picture.json")
        Call<String> getUserThumbUrl(@Path("id") String userId);

        @GET("/users/{id}.json")
        Flowable<User> getUserDetails(@Path("id") String userId);

        @GET("/users/{id}/events.json")
        Call<HashMap<String, EventInviteInfo>> getUsersEventInvites(@Path("id") String userId);

        @GET("/users/{id}/contacts.json")
        Flowable<HashMap<String, Boolean>> getUserContacts(@Path("id") String userId);

        @PUT("/users/{userId}.json")
        Call<User> addUser(@Body User user, @Path("userId") String userId);

        @PUT("/users/{userId}/events/{eventId}/isInviteRejected.json")
        Call<Boolean> rejectInvite(@Body Boolean bool, @Path("userId") String userId, @Path("eventId") String eventId);

        @PUT("/users/{userId}/events/{eventId}/isInviteAccepted.json")
        Call<Boolean> acceptInvite(@Body Boolean bool, @Path("userId") String userId, @Path("eventId") String eventId);

        @PUT("/events/{eventId}/users/{userId}.json")
        Call<Boolean> setEventUserAsAttending(@Body Boolean bool, @Path("userId") String userId, @Path("eventId") String eventId);


        @PATCH("/users/{userId}.json")
        Call<User> patchUser(@Body User user, @Path("userId") String userId);

        @PATCH("/users/{userId}/contacts.json")
        Call<HashMap<String, Boolean>> addContactToUser(@Body HashMap<String, Boolean> request, @Path("userId") String userId);

        @POST("/events.json")
        Call<EventCreationResponse> createEvent(@Body EventDetail request);

        //TODO: Clean this up. Find a way to minify the MiniEvent name
        @PUT("/users/{userId}/events/{eventId}.json")
        Call<EventInviteInfo> addEventToUser(@Body EventInviteInfo eventInviteInfo, @Path("userId") String userId, @Path("eventId") String eventId);


    }
}

File: app/src/main/java/com/bookyrself/bookyrself/services/FirebaseService.kt
package com.bookyrself.bookyrself.services

import com.bookyrself.bookyrself.data.serverModels.EventCreationResponse
import com.bookyrself.bookyrself.data.serverModels.EventDetail.EventDetail
import com.bookyrself.bookyrself.data.serverModels.EventDetail.Host
import com.bookyrself.bookyrself.data.serverModels.User.EventInviteInfo
import com.bookyrself.bookyrself.data.serverModels.User.User
import io.reactivex.Flowable
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.util.*

object FirebaseService {

    private const val BASE_URL_BOOKYRSELF_FIREBASE = "https://bookyrself-staging.firebaseio.com/"

    val instance: FirebaseApi by lazy {

        // Create the logging interceptor
        val interceptor = HttpLoggingInterceptor()
        interceptor.level = HttpLoggingInterceptor.Level.BODY
        val client = OkHttpClient.Builder().addInterceptor(interceptor).build()

        val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL_BOOKYRSELF_FIREBASE)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()

        retrofit.create(FirebaseApi::class.java)
    }

    interface FirebaseApi {

        @GET("/users/{id}.json")
        fun getUserDetails(@Path("id") userId: String): Flowable<User>

        @GET("/users/{id}/events.json")
        fun getUsersEventInvites(@Path("id") userId: String): Flowable<HashMap<String, EventInviteInfo>>

        @GET("/users/{id}/contacts.json")
        fun getUserContacts(@Path("id") userId: String): Flowable<HashMap<String, Boolean>>

        @PUT("/users/{userId}.json")
        fun updateUser(@Body user: User, @Path("userId") userId: String): Flowable<User>

        @PATCH("/users/{userId}.json")
        fun patchUser(@Body user: User, @Path("userId") userId: String): Flowable<User>

        @POST("/events.json")
        fun createEvent(@Body request: EventDetail): Flowable<EventCreationResponse>

        @PUT("/users/{userId}/events/{eventId}.json")
        fun addEventToUser(@Body eventInviteInfo: EventInviteInfo, @Path("userId") userId: String, @Path("eventId") eventId: String): Flowable<EventInviteInfo>

        @PUT("/events/{eventId}/host.json")
        fun updateEventHost(@Body host: Host, @Path("eventId") eventId: String): Flowable<Host>

        @PUT("/users/{userId}/unavailable_dates/{date}.json")
        fun setDateUnavailableForUser(@Body unavailable: Boolean?, @Path("userId") userId: String,
                                      @Path("date") date: String): Flowable<Boolean>

    }
}
